### ISSUE TITLE
cheriabitest failure: test_sandbox_system_calloc -- bounds exception in arena_avail_tree_remove()?

### ISSUE THREAD

**Issue by rwatson** [CTSRD-CHERI/cheribsd#143]
When I run the test ``test_sandbox_system_calloc``` in ```cheriabitest```, I get the following failure:
```
root@qemu-cheri:~ # cheriabitest test_sandbox_system_calloc
TEST: test_sandbox_system_calloc: Allocate memory in base for use in the sandbox
CHERI cause: ExcCode: 0x01 RegNum: $c01 (length violation)
$c00: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:1 s:0 p:7fff807d b:0000000160135000 l:0000000000004000 o:ffffffffffffb680 t:0
$c02: v:1 s:0 p:7fff807d b:00000001201618a8 l:0000000000000008 o:0 t:0
$c03: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:160200460 t:0
$c04: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:1604089c0 t:0
$c05: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:16041e000 t:0
$c06: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:16041e960 t:0
$c07: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c08: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c09: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c10: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c11: v:1 s:0 p:7fff807d b:0000000160135000 l:0000000000004000 o:0 t:0
$c12: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200dac60 t:0
$c13: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c14: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c15: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c16: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c17: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200da970 t:0
$c18: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:160200460 t:0
$c19: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:1604089c0 t:0
$c20: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:1602002c0 t:0
$c21: v:1 s:0 p:7fff807d b:00000001201618c0 l:0000000000000008 o:0 t:0
$c22: v:1 s:0 p:7fff807d b:00000001201618a8 l:0000000000000008 o:0 t:0
$c23: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:1604089c0 t:0
$c24: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c26: v:1 s:0 p:0000007d b:0000000160496000 l:00000000000001e0 o:0 t:0
$c31: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200dad80 t:0
FAIL: test_sandbox_system_calloc: Unexpected trusted stack unwind
root@qemu-cheri:~ # May 26 12:16:32 qemu-cheri kernel: USER_CHERI_EXCEPTION: pid 552 tid 100046 (cheriabitest), uid 0: CP2 fault (type 0x32)
May 26 12:16:32 qemu-cheri kernel: Trapframe Register Dump:
May 26 12:16:32 qemu-cheri kernel: zero: 0      at: 0x8000      v0: 0x12015ac60 v1: 0x5c80
May 26 12:16:32 qemu-cheri kernel: a0: 0x1201618a8      a1: 0   a2: 0   a3: 0x5
May 26 12:16:32 qemu-cheri kernel: a4: 0xffffffff       a5: 0x1201159c8 a6: 0x7fffffffffffffff  a6: 0x3f
May 26 12:16:32 qemu-cheri kernel: t0: 0x40     t1: 0x3 t2: 0xffffffffffffffff  t3: 0x5555555555555555
May 26 12:16:32 qemu-cheri kernel: t8: 0x3333333333333333       t9: 0x1200dac60 s0: 0   s1: 0xb9
May 26 12:16:32 qemu-cheri kernel: s2: 0x5      s3: 0   s4: 0x120156ee0 s5: 0x147
May 26 12:16:32 qemu-cheri kernel: s6: 0x8ac0   s7: 0x1100      k0: 0   k1: 0
May 26 12:16:32 qemu-cheri kernel: gp: 0x3680   sp: 0xffffffffffffb5e0  s8: 0xffffffffffffb5e0  ra: 0x18
May 26 12:16:32 qemu-cheri kernel: sr: 0x408084b3       mullo: 0x2840   mulhi: 0x5080   badvaddr: 0x1200dad80
May 26 12:16:32 qemu-cheri kernel: cause: 0x48  pc: 0x1200dad80
```

A bit of disassembly suggests a problem in ```jemalloc```'s ```arena_avail_tree_remove()```:
```
00000001200dac60 <arena_avail_tree_remove>:

        return (ret);
}

/* Generate red-black tree functions. */
rb_gen(static UNUSED, arena_avail_tree_, arena_avail_tree_t,
   1200dac60:   64010001        daddiu  at,zero,1
   1200dac64:   00010c3c        dsll32  at,at,0x10
   1200dac68:   6421ffff        daddiu  at,at,-1
   1200dac6c:   00010c38        dsll    at,at,0x10
   1200dac70:   64217e00        daddiu  at,at,32256
   1200dac74:   03a1e82d        daddu   sp,sp,at
   1200dac78:   340181f8        li      at,0x81f8
   1200dac7c:   03a1082d        daddu   at,sp,at
   1200dac80:   ebeb0803        csd     ra,at,0(c11)
   1200dac84:   340181f0        li      at,0x81f0
   1200dac88:   03a1082d        daddu   at,sp,at
   1200dac8c:   ebcb0803        csd     s8,at,0(c11)
   1200dac90:   340181e8        li      at,0x81e8
   1200dac94:   03a1082d        daddu   at,sp,at
   1200dac98:   eb8b0803        csd     gp,at,0(c11)
   1200dac9c:   340181e0        li      at,0x81e0
   1200daca0:   03a1082d        daddu   at,sp,at
   1200daca4:   eaeb0803        csd     s7,at,0(c11)
   1200daca8:   340181d8        li      at,0x81d8
   1200dacac:   03a1082d        daddu   at,sp,at
   1200dacb0:   eacb0803        csd     s6,at,0(c11)
   1200dacb4:   340181d0        li      at,0x81d0
   1200dacb8:   03a1082d        daddu   at,sp,at
   1200dacbc:   eaab0803        csd     s5,at,0(c11)
   1200dacc0:   340181c8        li      at,0x81c8
   1200dacc4:   03a1082d        daddu   at,sp,at
   1200dacc8:   ea8b0803        csd     s4,at,0(c11)
   1200daccc:   340181c0        li      at,0x81c0
   1200dacd0:   03a1082d        daddu   at,sp,at
   1200dacd4:   ea6b0803        csd     s3,at,0(c11)
   1200dacd8:   340181b8        li      at,0x81b8
   1200dacdc:   03a1082d        daddu   at,sp,at
   1200dace0:   ea4b0803        csd     s2,at,0(c11)
   1200dace4:   340181b0        li      at,0x81b0
   1200dace8:   03a1082d        daddu   at,sp,at
   1200dacec:   ea2b0803        csd     s1,at,0(c11)
   1200dacf0:   340181a8        li      at,0x81a8
   1200dacf4:   03a1082d        daddu   at,sp,at
   1200dacf8:   ea0b0803        csd     s0,at,0(c11)
   1200dacfc:   34018180        li      at,0x8180
   1200dad00:   03a1082d        daddu   at,sp,at
   1200dad04:   fb0b0800        csc     c24,at,0(c11)
   1200dad08:   34018160        li      at,0x8160
   1200dad0c:   03a1082d        daddu   at,sp,at
   1200dad10:   faeb0800        csc     c23,at,0(c11)
   1200dad14:   34018140        li      at,0x8140
   1200dad18:   03a1082d        daddu   at,sp,at
   1200dad1c:   facb0800        csc     c22,at,0(c11)
   1200dad20:   34018120        li      at,0x8120
   1200dad24:   03a1082d        daddu   at,sp,at
   1200dad28:   faab0800        csc     c21,at,0(c11)
   1200dad2c:   34018100        li      at,0x8100
   1200dad30:   03a1082d        daddu   at,sp,at
   1200dad34:   fa8b0800        csc     c20,at,0(c11)
   1200dad38:   340180e0        li      at,0x80e0
   1200dad3c:   03a1082d        daddu   at,sp,at
   1200dad40:   fa6b0800        csc     c19,at,0(c11)
   1200dad44:   340180c0        li      at,0x80c0
   1200dad48:   03a1082d        daddu   at,sp,at
   1200dad4c:   fa4b0800        csc     c18,at,0(c11)
   1200dad50:   341c80a0        li      gp,0x80a0
   1200dad54:   03bce02d        daddu   gp,sp,gp
   1200dad58:   fa2be000        csc     c17,gp,0(c11)
   1200dad5c:   03a0f025        move    s8,sp
   1200dad60:   49b32000        cmove   c19,c4
   1200dad64:   49b96002        cgetoffset      t9,c12
   1200dad68:   3c010008        lui     at,0x8
   1200dad6c:   0039102d        daddu   v0,at,t9
   1200dad70:   67c100a0        daddiu  at,s8,160
   1200dad74:   dae30000        clc     c23,zero,0(c3)
   1200dad78:   49a15841        csetoffset      c1,c11,at
   1200dad7c:   34018000        li      at,0x8000
   1200dad80:   48270840        csetbounds      c7,c1,at         <-- Boom here
   1200dad84:   fae70000        csc     c23,zero,0(c7)
   1200dad88:   64010020        daddiu  at,zero,32
   1200dad8c:   49b51840        cincoffset      c21,c3,at
   1200dad90:   49c1bd40        ceq     at,c23,c21
   1200dad94:   1420008c        bnez    at,1200dafc8 <arena_avail_tree_remove+0x
368>
   1200dad98:   48850007        cfromptr        c5,c0,zero
   1200dad9c:   f8abf002        csc     c5,s8,32(c11)
   1200dada0:   f86bf006        csc     c3,s8,96(c11)
   1200dada4:   6441c280        daddiu  at,v0,-15744
```

The code appears to be trying to set the bounds on ```$c12``` to 0x8000, but its starting bounds are already 0x4000. The caller (from ```$c17```) appears to be:
```
        assert(npages == (arena_mapbits_unallocated_size_get(chunk, pageind) >>
            LG_PAGE));
        arena_avail_tree_remove(&arena->runs_avail, arena_miscelm_get(chunk,
   1200da928:   640201a0        daddiu  v0,zero,416
   1200da92c:   49b2a080        cincoffset      c18,c20,v0
{

        assert(pageind >= map_bias);
        assert(pageind < chunk_npages);

        return ((arena_chunk_map_misc_t *)((uintptr_t)chunk +
   1200da930:   49a198c0        cincoffset      c1,c19,v1
            (uintptr_t)map_misc_offset) + pageind-map_bias);
   1200da934:   001111b8        dsll    v0,s1,0x6
   1200da938:   001119f8        dsll    v1,s1,0x7
   1200da93c:   0062b02d        daddu   s6,v1,v0
   1200da940:   49a10d80        cincoffset      c1,c1,s6
   1200da944:   0001102f        dnegu   v0,at
   1200da948:   000109b8        dsll    at,at,0x6
   1200da94c:   000211f8        dsll    v0,v0,0x7
   1200da950:   0041082f        dsubu   at,v0,at
   1200da954:   49a40840        cincoffset      c4,c1,at
   1200da958:   3c010000        lui     at,0x0
   1200da95c:   0034082d        daddu   at,at,s4
   1200da960:   dc21ad80        ld      at,-21120(at)
   1200da964:   480c09ff        cgetpccsetoffset        c12,at
   1200da968:   48f16000        cjalr   c17,c12
   1200da96c:   49a39000        cmove   c3,c18
            flag_dirty);
        assert(need_pages <= total_pages);
        rem_pages = total_pages - need_pages;

        arena_avail_remove(arena, chunk, run_ind, total_pages);
        if (flag_dirty != 0)
```

**Comment by rwatson** [comment]
(@davidchisnall might take an interest in this -- ```$c1``` is derived from the stack pointer, and so likely a stack allocation generated by the compiler...?)

**Comment by rwatson** [comment]
Closed as resolved by 25526e6.
